be1-go/channel/register/mod.go

```diff
@@ -0,0 +1,43 @@
+package register
+
+import "popstellar/message/query/method/message"
+
+type MessageRegistry struct {
+	Registry map[string]CallbackData
+}
+
+type CallbackData struct {
+	Callback     func(message.Message, interface{}) error
+	ConcreteType interface{}
+}
+
+func NewMessageRegistry() MessageRegistry {
+	return MessageRegistry{
+		Registry: make(map[string]CallbackData),
+	}
+}
+
+// register registers a new action that will be associated to a callback
+// function. c defines the concrete type the callback function will receive. c
+// must NOT be a pointer. The callback function will have to cast the interface
+// it receives to the concrete type. For example:
+//
+//   // define a callback
+//   func execElect(msg interface) error {
+//	   m, ok := msg.(*messagedata.ConsensusElect)
+//     if !ok {...}
+//     // ...
+//   }
+//
+//   // register the "elect" type
+//   registry.register("elect", execElect, messagedata.ConsensusElect{})
+//
+//   // when we need to process a message we call "processMsg"
+//   err = registry.processMsg(msg)
+//
+func (m MessageRegistry) Register(action string, f func(message.Message, interface{}) error, c interface{}) {
+	m.Registry[action] = CallbackData{
```

Comment:
In fact we could simplify it and take the "action" from `c` directly. You do that by having a map of `reflect.Type`:

```go
map[reflect.Type]
```

Then you can use:

```go
key := reflect.TypeOf(c)
map[key] = CallbackData{...}
```

Reply:
And add a function:

```go
func (m MessageRegistry) Process(c interface{}) error {
  key := reflect.TypeOf(c)

  data, found := m.Registry[key]
  if !found {...}

  concreteType := reflect.New(reflect.ValueOf(data.ConcreteType).Type()).Interface()

	err := msg.UnmarshalData(&concreteType)
	if err != nil {
		return xerrors.Errorf("failed to unmarshal data: %v", err)
	}

	err = data.Callback(c, concreteType)
	if err != nil {
		return xerrors.Errorf("failed to call message '%s': %v", key, err)
	}

  return nil
}

Reply:
Then the attribute of `MessageRegistry` (`Registry`) can be private.